cupsfilters/rastertops.c

```diff
@@ -0,0 +1,527 @@
+		
+  /* allocate deflate state */
+  strm.zalloc = Z_NULL;
+  strm.zfree = Z_NULL;
+  strm.opaque = Z_NULL;
+  ret = deflateInit(&strm, -1);
+  if (ret != Z_OK)
+    return ret;
+
+  if(header.cupsBitsPerColor == 1 &&
+     (header.cupsColorSpace == CUPS_CSPACE_RGB ||
+      header.cupsColorSpace == CUPS_CSPACE_ADOBERGB ||
+      header.cupsColorSpace == CUPS_CSPACE_SRGB))
+    flag = 1;
+
+  /* compress until end of file */
+  do {
+    pixdata = malloc(header.cupsBytesPerLine);
+    cupsRasterReadPixels(ras, pixdata, header.cupsBytesPerLine);
+    if (flag)
+    {
+      convertedpix = malloc(header.cupsBytesPerLine * 6);
+      convert_pixels(pixdata,convertedpix, header.cupsBytesPerLine);
+      alloc = header.cupsBytesPerLine * 6;
+    }
+    else
+    {
+      convertedpix = malloc(header.cupsBytesPerLine);
+      memcpy(convertedpix, pixdata, header.cupsBytesPerLine);
+      alloc = header.cupsBytesPerLine;
+    }
+
+    if(curr_line == header.cupsHeight)
+      flush = Z_FINISH;
+    else
+      flush = Z_NO_FLUSH;
+    curr_line++;
+    memcpy(in, convertedpix, alloc);
+    strm.avail_in = alloc;
+    strm.next_in = in;
+
+    /* run deflate() on input until output buffer not full, finish
+     * compression if all of source has been read in */
+    do {
+      strm.avail_out = alloc;
+      strm.next_out = out;
+
+      /* Run the deflate algorithm on the data */
+      ret = deflate(&strm, flush);
+
+      /* check whether state is not clobbered */
+      assert(ret != Z_STREAM_ERROR);
+      have = alloc - strm.avail_out;
+      if (fwrite(out, 1, have, doc->outputfp) != have)
+      {
+	(void)deflateEnd(&strm);
+	if (convertedpix != NULL)
+	  free(convertedpix);
+	return Z_ERRNO;
+      }
+    } while (strm.avail_out == 0);
+
+    /* all input will be used */
+    assert(strm.avail_in == 0);
+
+    /* done when last data in file processed */
+    free(pixdata);
+    free(convertedpix);
+  } while (flush != Z_FINISH);
+
+  /* stream will be complete */
+  assert(ret == Z_STREAM_END);
+
+  /* clean up and return */
+  (void)deflateEnd(&strm);
+  return Z_OK;
+}
+
+/*
+ *  Report a zlib or i/o error
+ */
+
+void
+zerr(int ret, /* I - Return status of deflate */
+    rastertops_doc_t *doc) /* I - Document information */
+{
+  if (doc->logfunc) doc->logfunc(doc->logdata, FILTER_LOGLEVEL_ERROR,
+		   "zpipe: ");
+  switch (ret) {
+  case Z_ERRNO:
+    if (doc->logfunc) doc->logfunc(doc->logdata, FILTER_LOGLEVEL_ERROR,
+		  "error in source data or output file\n");
+    break;
+  case Z_STREAM_ERROR:
+    if (doc->logfunc) doc->logfunc(doc->logdata, FILTER_LOGLEVEL_ERROR,
+		  "invalid compression level\n");
+    break;
+  case Z_DATA_ERROR:
+    if (doc->logfunc) doc->logfunc(doc->logdata, FILTER_LOGLEVEL_ERROR,
+		  "invalid or incomplete deflate data\n");
+    break;
+  case Z_MEM_ERROR:
+    if (doc->logfunc) doc->logfunc(doc->logdata, FILTER_LOGLEVEL_ERROR,
+		  "out of memory\n");
```

Comment:
When logging, please start the log messages with "rastertops: ", for example here do
```
if (doc->logfunc) doc->logfunc(doc->logdata, FILTER_LOGLEVEL_ERROR,
￼		  "rastertops: out of memory\n");
```
This makes it easier to see from which filter function a log message comes, especially when calling a chain of filters.
See also the other filter functions.
Could you correct this with all log messages? Thanks.